### PR TITLE
Use GET and preload links on Server Islands

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- Makes it so that Server Islands use a `GET` request when they can, and include a `<link rel="preload">` tag to kick off the request faster.

## Testing

Deployed to server-islands.com and ran the tests a few times. Here are the relevant waterfalls:

- [Using GET](https://www.webpagetest.org/result/240815_AiDc6J_B2N/1/details/#waterfall_view_step1)
- [Using POST](https://www.webpagetest.org/result/240716_AiDc6D_D1N/1/details/#waterfall_view_step1)

### GET

Screen Shot 2024-08-15 at 2 49 38 PM

### POST

Screen Shot 2024-08-15 at 2 49 45 PM

### Results

Using GET with the preload is causing these requests to start between 180 - 400ms faster. Looking at the waterfalls I think the biggest factor here is that the preloads are being kicked off before images. Whereas with POST the requests are competing with the images. This might be a case where this particular app has a larger effect than others will.

## Docs

N/A

### CODE REVIEW

**Comment by matthewp** [comment]
!preview si-get

**Comment by matthewp** [comment]
!preview si-get

**Comment by matthewp** [comment]
!preview si-get

**Comment by matthewp** [comment]
!preview si-get

File: packages/astro/src/core/server-islands/endpoint.ts [hunk 1]
```diff
@@ -53,12 +53,52 @@ type RenderOptions = {
 	slots: Record<string, string>;
 };
 
+function badRequest() {
```
[the remark this answers is not in the document]
Comment (matthewp) [reply packages/astro/src/core/server-islands/endpoint.ts:56]: Yeah, I started to do this but thought it was giving too much information to a potential attacker. But looking at https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400 it seems common to give a reason. So I can add this.

File: packages/astro/src/core/server-islands/endpoint.ts [hunk 2]
```diff
@@ -53,12 +53,52 @@ type RenderOptions = {
...
+			try {
+			const raw = await request.text();
+			const data = JSON.parse(raw) as RenderOptions;
+			return data;
+			} catch {
+				return badRequest();
+			}
+		}
+		default: {
+			return badRequest();
```
[the remark this answers is not in the document]
Comment (matthewp) [reply packages/astro/src/core/server-islands/endpoint.ts]: done

File: packages/astro/src/core/server-islands/endpoint.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (matthewp) [reply packages/astro/src/core/server-islands/endpoint.ts]: done

**Review by ematipico — APPROVED** [review]

**Pull request merged**